### DISCUSSION TITLE
How to set the Callback URL via URL Query?

### DISCUSSION THREAD

**Discussion by lamualfa** [nextauthjs/next-auth#2508] [Help]
I have a Telegram bot and user can sign in via my Next.js app.

The flow is user send a message to my telegram bot, telegram bot will send the Sign In URL. Then, user will visit the Sign In URL that has been sent by my telegram bot. After user sign in with the given URL, user will redirect to custom page that contain next instruction about next step.

The question is, how to generate the sign in link with custom callback URL programmatically via Server that can send to another user?

I have tried passing the url of the custom page in `callbackUrl` query, but it's not work. The user is even directed to the `/` page after signing in.

`http://localhost:3000/api/auth/signin?callbackUrl=http://localhost:3000/instruction/telegram`

Thanks.

**Comment by sastraxi** [comment]
If you are OK targeting browsers that support [localstorage](https://caniuse.com/mdn-api_window_localstorage), I might suggest this:

1. Generate your sign-in link with some querystring parameter, e.g. `?next_step=...`
2. When your component mounts, extract the `next_step` query parameter [on the frontend](https://stackoverflow.com/a/901144/220642), then store it using localstorage
4. Allow the user to sign-in normally, after which they will be redirected to the default post-sign-in page
5. On that page, check localstorage to see if `next_step` has been set. If so, remove it from local storage and navigate the user there, e.g. using [next/router](https://nextjs.org/docs/api-reference/next/router)

This assumes you "control" the page the user is first landing on. If it's a next-auth `/api` route, you can always redirect to that afterwards from a custom sign-in URL.

**Reply by lamualfa** [reply]
Does that mean we cannot passing or setting `callbackUrl` programmatically without `signIn` function from Next Auth library?

**Reply by sastraxi** [reply]
Unfortunately I'm a little out of my depth here (just learning the library myself) and am not sure. However, if the callback URL you're generating will ultimately be passed to the (e.g. OAuth) provider, you'll also need to make sure it's whitelisted on their end. Generating callback URLs programatically in auth flows is pretty tricky for this reason, so I'd suggest an approach like the above even without knowing too many details.
